be/src/olap/rowset/segment_v2/vertical_segment_writer.cpp

```diff
@@ -689,8 +689,11 @@ Status VerticalSegmentWriter::batch_block(const vectorized::Block* block, size_t
         _opts.rowset_ctx->partial_update_info->is_partial_update &&
         _opts.write_type == DataWriteType::TYPE_DIRECT &&
         !_opts.rowset_ctx->is_transient_rowset_writer) {
-        if (block->columns() <= _tablet_schema->num_key_columns() ||
-            block->columns() >= _tablet_schema->num_columns()) {
+        if ((_tablet_schema->keys_type() == UNIQUE_KEYS &&
```

Comment:
you should also consider SegmentWriter class, it has same method and code with VerticalSegmentWriter